File: 1806.minimum-number-of-operations-to-reinitialize-a-permutation.cpp
/*
 * @lc app=leetcode id=1806 lang=cpp
 *
 * [1806] Minimum Number of Operations to Reinitialize a Permutation
 */

// @lc code=start
#include <vector>
#include <deque>
#include <algorithm>

class Solution {
public:
    /*
    We can use BFS from each index and find all the possible positions. The result is the maximum cluster size.
    */
    int reinitializePermutation(int n) {
        std::vector<bool> seen(n, false);
        int result = 0;
        
        for (int i = 0; i < n; i++) {
            if (!seen[i]) {
                std::deque<int> dq;
                seen[i] = true;
                dq.push_back(i);
                int ans = 0;
                
                while (!dq.empty()) {
                    int node = dq.front();
                    dq.pop_front();
                    ans++;
                    int nxt = (node % 2 == 0) ? node / 2 : n / 2 + (node - 1) / 2;
                    
                    if (!seen[nxt]) {
                        seen[nxt] = true;
                        dq.push_back(nxt);
                    }
                }
                
                result = std::max(result, ans);
            }
        }
        
        return result;
    }
};

// @lc code=end

File: 1806.minimum-number-of-operations-to-reinitialize-a-permutation.py
#
# @lc app=leetcode id=1806 lang=python3
#
# [1806] Minimum Number of Operations to Reinitialize a Permutation
#

# @lc code=start
class Solution:
    '''
    We can use BFS from each index and find all the possible positions. The result the maximum cluster size
    '''
    def reinitializePermutation(self, n: int) -> int:
        seen = [False] * n
        result = 0
        for i in range(n):
            if not seen[i]:
                dq = deque()
                seen[i] = True
                dq.append(i)
                ans = 0
                while dq:
                    node = dq.popleft()
                    ans += 1
                    nxt = node // 2 if node % 2 == 0 else n // 2 + (node - 1) // 2
                    if not seen[nxt]:
                        seen[nxt] = True
                        dq.append(nxt)
                result = max(result, ans)
        return result
# @lc code=end

